src/main/java/uk/gov/hmcts/dts/fact/services/admin/AdminCourtAddressService.java

```diff
@@ -55,7 +65,57 @@ public AdminCourtAddressService(final CourtRepository courtRepository, final Cou
             .collect(toList());
     }
 
-    @SuppressWarnings("PMD.DataflowAnomalyAnalysis")
+    public List<String> validateCourtAddressPostcodes(final String slug, final List<CourtAddress> courtAddresses) {
+        final List<String> invalidPostcodes = new ArrayList<>();
+
+        if (!CollectionUtils.isEmpty(courtAddresses)) {
+            final List<String> allPostcodes = getAllPostcodesSortedByAddressType(courtAddresses);
+
+            if (!CollectionUtils.isEmpty(allPostcodes)) {
+                invalidPostcodes.addAll(validationService.validateFullPostcodes(allPostcodes));
+            }
+
+            // Only update the in-person court coordinates is all input postcodes are valid
+            if (invalidPostcodes.isEmpty()
+                && !CollectionUtils.isEmpty(allPostcodes)
+                && isInPersonCourt(slug)) {
+                updateCourtLatLonUsingPrimaryPostcode(slug, allPostcodes.get(0));
+            }
+        }
+        return invalidPostcodes;
+    }
+
+    private List<String> getAllPostcodesSortedByAddressType(final List<CourtAddress> courtAddresses) {
+        final Map<Integer, uk.gov.hmcts.dts.fact.entity.AddressType> addressTypeMap = addressTypeService.getAddressTypeMap();
+
+        // Sort the postcodes so that the 'visit us' or 'visit or contact us' postcodes appear first
+        return courtAddresses.stream()
+            .sorted(Comparator.comparingInt(a -> AddressType.isCourtAddress(getAddressTypeFromId(addressTypeMap, a.getAddressTypeId())) ? 0 : 1))
+            .map(CourtAddress::getPostcode)
+            .filter(StringUtils::isNotBlank)
+            .collect(toList());
+    }
+
+    private String getAddressTypeFromId(final Map<Integer, uk.gov.hmcts.dts.fact.entity.AddressType> map, final Integer addressTypeId) {
+        if (!map.containsKey(addressTypeId)) {
+            throw new IllegalArgumentException("Unknown address type ID: " + addressTypeId);
+        }
+        return map.get(addressTypeId).getName();
+    }
+
+    private boolean isInPersonCourt(final String slug) {
+        final Court courtEntity = courtRepository.findBySlug(slug)
+            .orElseThrow(() -> new NotFoundException(slug));
+        return courtEntity.getInPerson() == null || courtEntity.getInPerson().getIsInPerson();
+    }
+
+    private void updateCourtLatLonUsingPrimaryPostcode(final String slug, final String postcode) {
+        final Optional<MapitData> mapitData = mapitService.getMapitData(postcode);
+        if (mapitData.isPresent()) {
+            adminService.updateCourtLatLon(slug, mapitData.get().getLat(), mapitData.get().getLon());
```

Comment:
Just double checking on this one (I think it's all good) but the mapit data based on a postcode search may have information related to a town, or some location, that we would be able to cross reference to the rest of the address. There is no requirement or need for us to do that at the moment, right? :- )

Reply:
Yeah there is not required to do that. I know the call to get the mapit area is available but that call specifically passes in the local authority types. We will need to do something slightly different to determine the town/city type. We validate the postcodes because we need that to set the coordinates. It is better if we can validate the entire address but it is not possible at the moment